REGRESSION(291808@main): [macOS Debug]: ASSERTION FAILED: !index in CSSFontSelector::fallbackFontAt on html5lib/generated/run-entities01-data.html (290012)

vitorroriz opened the pull request:
#### 0d680136cb51cf449dca3430455b1ffe95be5661
<pre>
REGRESSION(291808@main): [macOS Debug]: ASSERTION FAILED: !index in CSSFontSelector::fallbackFontAt on html5lib/generated/run-entities01-data.html (290012)
<a href="https://rdar.apple.com/147365119">rdar://147365119</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=290231">https://bugs.webkit.org/show_bug.cgi?id=290231</a>

Reviewed by Geoffrey Garen.

Settings control how we resolve generic font-family names.

This bug happens because we were setting setFontFallbackPrefersPictographs(true)
from another test, on the same process, before running run-entities01-data.html.

This made the cached FontCascadeFont object to have its m_lastRealizedFallbackIndex
expanded beyound effectiveFamilyCount(), making fontSelectorFallbackIndex turning 1.

When running run-entities01-data.html, m_lastRealizedFallbackIndex is still
retaining that value. However, fontSelector-&gt;fallbackFontCount() that comes from Settings
is now 0 and we won&apos;t pass the check `fontSelectorFallbackIndex == fontSelector-&gt;fallbackFontCount()`.

This lead us to a inconsistent state in which we call fallbackFontAt for resolving the pictographFontFamily,
although we are stating by the index != 0 that we already did.

We should make sure that setting fontFallbackPrefersPictographs invalidates the FontCascadeCache
as we already do for the other related settings as `setSansSerifFontFamily`, `setCursiveFontFamily` and etc.

* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/page/Settings.yaml:
* Source/WebCore/page/SettingsBase.cpp:
(WebCore::SettingsBase::protectedPage const):
(WebCore::SettingsBase::fontFallbackPrefersPictographsChanged):
* Source/WebCore/page/SettingsBase.h:

Canonical link: <a href="https://commits.webkit.org/292764@main">https://commits.webkit.org/292764@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9342edfbd7065732dd6ddde067772553d14fa896

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96984 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16598 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/6800 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102060 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47505 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99029 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16894 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25047 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73884 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31092 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99987 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12740 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/87726 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54220 "Found 1 new API test failure: /WPE/TestSSL:/webkit/WebKitWebView/ephemeral-tls-errors (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12495 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5565 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46832 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82574 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5644 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104081 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24052 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17554 "Found 1 new test failure: imported/w3c/web-platform-tests/workers/abrupt-completion.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82930 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/24305 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83817 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82325 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20717 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27007 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4555 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/17555 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24014 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29248 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/23841 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27153 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25414 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebCore/platform/graphics/FontGenericFamilies.cpp:
```diff
@@ -47,6 +48,15 @@ static bool setGenericFontFamilyForScript(ScriptFontFamilyMap& fontMap, const St
+bool FontGenericFamilies::setGenericFontFamilyForScriptAndComputeHash(ScriptFontFamilyMap& fontMap, const String& family, UScriptCode script, WebKitFontFamilyNames::FamilyNamesIndex genericFamily)
+{
+    auto setResult = setGenericFontFamilyForScript(fontMap, family, script);
+        // We just care about the common script because API only allows setting the resolution for this script. Therefore, that's the only part that can be mutated by Settings. We care about common script even when the content's script is not common because it is used as a fallback if we can't resolve with the original script (genericFontFamilyForScript).
```
What technical mechanism guarantees that this property of API will remain true going forward?

I don't see such a mechanism, so I think this approach is too brittle.

One approach that might be less brittle is to give the Settings object a generation counter, and bump the generation when Settings change. Then make the generation counter part of the cache key. Settings change rarely so this I likely to be affordable. And if all Settings setters bump the generation counter, hopefully that's a sufficient indicator to future authors to maintain that behavior.

Another approach is for the Settings setter that invalidates the font cache to just clear the cache explicitly. Here again, hopefully putting code like that inside Settings will be a clear indicator to future authors of Settings changes.

vitorroriz replied:
The idea is that every time one of the relevant setters (the ones setting resolved names for generic families) are called, we recompute the hash for the list of font families. I think this is similar to generation bump, but instead of having a incremental counter we calculate the hash for the list of names directly, so we can use it in FontCascadeCache key.

geoffreygaren replied:
Sure, but what technical mechanism ensures the correctness of this comment about the `if (script == USCRIPT_COMMON)` check? Somehow we need to simplify or clarify.

And isn't `invalidateAfterGenericFamilyChange` the place we do this kind of invalidation today?

vitorroriz replied:
I think I got your point. The calls from the API itself will always result in `USCRIPT_COMMON`, as Script is not exposed and we always pass script common for these cases. However, you are saying that technically, something on WebCore could still set it for another script, right?

I think we can use the generation counter as you mentioned but as part specifically of FontGenericFamilies as we don't care for other Settings here.

cdumez commented on Source/WebCore/page/SettingsBase.h:
```diff
@@ -166,8 +166,10 @@ class SettingsBase : public AbstractRefCounted {
+    void fontFallbackPrefersPictographsChanged();
 
     WeakPtr<Page> m_page;
+    RefPtr<Page> protectedPage() const;
```
Please do not add member functions in the middle of data members. This is definitely NOT the pattern in WebKit code.

vitorroriz committed:
REGRESSION(291808@main): [macOS Debug]: ASSERTION FAILED: !index in CSSFontSelector::fallbackFontAt on html5lib/generated/run-entities01-data.html (290012)
rdar://147365119
https://bugs.webkit.org/show_bug.cgi?id=290231

Reviewed by Geoffrey Garen.

Settings control how we resolve generic font-family names.

This bug happens because we were setting setFontFallbackPrefersPictographs(true)
from another test, on the same process, before running run-entities01-data.html.

This made the cached FontCascadeFont object to have its m_lastRealizedFallbackIndex
expanded beyound effectiveFamilyCount(), making fontSelectorFallbackIndex turning 1.

When running run-entities01-data.html, m_lastRealizedFallbackIndex is still
retaining that value. However, fontSelector->fallbackFontCount() that comes from Settings
is now 0 and we won't pass the check `fontSelectorFallbackIndex == fontSelector->fallbackFontCount()`.

This lead us to a inconsistent state in which we call fallbackFontAt for resolving the pictographFontFamily,
although we are stating by the index != 0 that we already did.

We should make sure that setting fontFallbackPrefersPictographs invalidates the FontCascadeCache
as we already do for the other related settings as `setSansSerifFontFamily`, `setCursiveFontFamily` and etc.

* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/page/Settings.yaml:
* Source/WebCore/page/SettingsBase.cpp:
(WebCore::SettingsBase::protectedPage const):
(WebCore::SettingsBase::fontFallbackPrefersPictographsChanged):
* Source/WebCore/page/SettingsBase.h:

Canonical link: https://commits.webkit.org/292764@main
